### PR TITLE
Move tech design overview from wiki into repo

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
The technical design overview document currently on the wiki should be added to the NVDA repository so it can reflect the state of the commit it is read in.
 
### Description of how this pull request fixes the issue:
Coped the technical design overview from the NVDA wiki to devDocs/technicalDesignOverview.md. Reformatted to break lines between sentences.

### Testing strategy:
None

### Known issues with pull request:
Should continue to review the document and ensure it is up to date.

### Change log entries:
New features
Changes
Bug fixes
For Developers

### Code Review Checklist:

- [ ] Pull Request description is up to date.
- [ ] Unit tests.
- [ ] System (end to end) tests.
- [ ] Manual testing.
- [ ] User Documentation.
- [ ] Change log entry.
- [ ] Context sensitive help for GUI changes.
- [ ] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### CODE REVIEW

**Comment by CyrilleB79** [comment]
Maybe a point should be added in the "Code Review Checklist" item of the PR template:
`* [] Developer Documentation`

Or generalize the "User documentation" item as follows:
``* [] Documentation (user documentation, developer documentation, other?)`

File: devDocs/technicalDesignOverview.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+Aside from application specific customisation using [app modules](#app-modules), it is also possible to extend NVDA on a global level.
+For example, new global commands can be added, behaviour can be changed and new GUI toolkits can be supported.
+This can be done using global plugins.
+A global plugin is derived from the `globalPluginHandler.GlobalPlugin` base class.
+Similar to [global commands](#global-commands), they can bind scripts which can be executed everywhere.
+They can also implement their own global [NVDA Objects](#nvda-objects).
+
+### Tree Interceptors
+Sometimes, it is necessary to intercept events and scripts for an entire hierarchy (or tree) of [NVDA objects](#nvda-objects).
+For example, this is necessary to seemlessly handle complex documents which consist of many objects.
```
Comment (lukaszgo1) [inline devDocs/technicalDesignOverview.md:155]: seemlessly -> seamlessly

**Review by lukaszgo1 — CHANGES_REQUESTED** [review]

**Comment by lukaszgo1** [comment]
Since we're  moving many pages from the wiki to the repository has it been considered to remove them from the wiki? This would not only avoid duplication of the content but also ensure that there is only one source of information which needs to be kept uptodate.

File: devDocs/technicalDesignOverview.md [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+ 
+ You can fetch a TextInfo object from an NVDA object via its `makeTextInfo` method, passing in the particular `textInfos.POSITION_*` constant depending on whether you want to fetch a range representing the position of the caret, selection, start or end of the text, or the entire text.
+ 
+### Global Commands
+The global commands object (`globalCommands.GlobalCommands`) contains built-in global scripts; i.e.
+they can be executed everywhere.
+For example, the review, report current focus and date/time scripts are all located in global commands.
+
+### Plugins
+NvDA allows third-parties to extend NvDA's functionality through plugins and add-ons.
```
Comment (seanbudd) [inline devDocs/technicalDesignOverview.md:143]:
```suggestion
NVDA allows third-parties to extend NVDA's functionality through plugins and add-ons.
```

File: devDocs/technicalDesignOverview.md [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+A tree interceptor (TreeInterceptor) is derived from the `treeInterceptorHandler.TreeInterceptor` base class.
+It receives events and scripts for all [NVDA objects](#nvda-objects) beneath and including the root NVDA object of the tree interceptor.
+Tree interceptors are created when a TreeInterceptor class is returned from the `treeInterceptorClass` property of an NVDA object.
+Tree interceptors are used mostly for web documents, where all events and scripts for NVDA objects within a document need to be handled by the document (root NVDA object) itself.
+
+#### Browse mode documents
+Complex documents such as web pages are very often not flat; i.e. information does not simply run from top to bottom.
+Because of this, complex document browsers often do not provide a way to navigate documents using the caret, and even when they do, it is often problematic.
+Therefore, screen readers need to create their own flat representation of a document from the object hierarchy provided by the browser and allow the user to navigate this flat representation.
+Browse mode documents are a subclass of `TreeInterceptor` that provide scripts that allow navigating the document in a linea fassion.
```
Comment (seanbudd) [inline devDocs/technicalDesignOverview.md:177]:
```suggestion
Browse mode documents are a subclass of `TreeInterceptor` that provide scripts that allow navigating the document in a linear fashion.
```

File: devDocs/technicalDesignOverview.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+Most events do not take any additional arguments.
+Global plugins, app modules and tree interceptors are passed a handler function which should be called if the event should be handled by the next handler;
+e.g. the object itself.
+
+Although an event is always for a particular NVDA object, it first has a chance of being handled by global plugins, app modules or tree interceptors.
+If an event is handled by one of these, meaning that an `event_*` method was found and executed, the event stops there and does not go further, unless the method that handled it specifically calls the `nextHandler` function object passed to it. 
+
+The chain of handlers is as follows:
+* The first found global plugin
+* The next found global plugin...
```
Comment (seanbudd) [inline devDocs/technicalDesignOverview.md:207]:
```suggestion
* The next found global plugin (until no more are found)
```

File: devDocs/technicalDesignOverview.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+Input gestures are bound to scripts in the class using a `scriptHandler.script` function decorator.
+They can also be bound at runtime using `bindGesture`.
+These are inherited from `baseObject.ScriptableObject`.
+
+Similar to events, input gestures have a chance to be handled by a script at one of many levels.
+But unlike events, once an input gesture finds and executes a script, there is no clean way to have the input gesture handled by a subsequent level.
+ 
+ The chain of handlers is as follows:
+* The first found global plugin 
+* The next found global plugin...
```
Comment (seanbudd) [inline devDocs/technicalDesignOverview.md:225]:
```suggestion
* The next found global plugin (until no more are found)
```

**Review by seanbudd — APPROVED** [review]
LGTM. I've noted a couple of typos, and a clarification on what is meant by (...)

File: devDocs/technicalDesignOverview.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+Aside from application specific customisation using [app modules](#app-modules), it is also possible to extend NVDA on a global level.
+For example, new global commands can be added, behaviour can be changed and new GUI toolkits can be supported.
+This can be done using global plugins.
+A global plugin is derived from the `globalPluginHandler.GlobalPlugin` base class.
+Similar to [global commands](#global-commands), they can bind scripts which can be executed everywhere.
+More specifically, global plugins receive events for all [NVDA objects](#nvda-objects) in the Operating System and can bind scripts which can be executed anywhere.
+They can also implement their own global [NVDA Objects](#nvda-objects).
+
+### Tree Interceptors
+Sometimes, it is necessary to intercept events and scripts for an entire hierarchy (or tree) of [NVDA objects](#nvda-objects).
```
Comment (feerrenrut) [inline devDocs/technicalDesignOverview.md:165]: This leaves me wondering, intercept between which two points? Where is the event "coming from", and where is it "going"?

File: devDocs/technicalDesignOverview.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+* The tree interceptor containing the currently focused NVDA object, I.e. fetched from the NVDA object's `treeInterceptor` property if the property is not `None`
+* The currently focused NVDA object
+* the first ancestor (parent) of the currently focused NVDAObject, if the found script's `canPropagate` property is True
+* the next ancestor of the currently focused NVDAObject, if the found script's `canPropagate` property is True...
+ * Global commands
+ 
+## Inter-process Communication
+In general terms, every running application or service on a computer, including NVDA, is a separate process.
+No process can access data in another process except via special operating system mechanisms.
+This is called inter-process communication (IPC).
```
Comment (feerrenrut) [inline devDocs/technicalDesignOverview.md:236]: It would be helpful if the specific IPC approaches used by NVDA were described.

**Pull request merged**